x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -355,7 +446,7 @@ private StackTraceHandler(
         ) {
             this.clusterState = clusterState;
             this.stackTracePerId = new ConcurrentHashMap<>(stackTraceCount);
-            this.remainingSlices = new AtomicInteger(slices);
+            this.expectedSlices = new AtomicInteger(slices + 1); // +1 for onHostsResponse
```

Comment:
This should be passed from the outside. The handler should not have knowledge about the surrounding context.